File: app/src/main/java/com/bearcave/passageplanning/base/BaseManagerAdapter.java
package com.bearcave.passageplanning.base;

import android.content.Context;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.bearcave.passageplanning.R;
import com.bearcave.passageplanning.data.database.tables.base.CRUD;

import java.util.ArrayList;

import butterknife.ButterKnife;

/**
 * Created by miwas on 20.05.17.
 */

public abstract class BaseManagerAdapter<T extends OnNameRequestedListener> extends BaseExpandableListAdapter {

    private ArrayList<T> container;
    private Context context;
    private LayoutInflater inflater;
    private final SparseArray<OnMenuItemSelectedAction> menuActions = new SparseArray<>();
    private final CRUD<T> database;


    public BaseManagerAdapter(BaseManagerFragment parent, Context context) {
        this.context = context;
        inflater = LayoutInflater.from(context);
        database = (CRUD<T>) parent;

        menuActions.put(EDIT_CODE, t -> database.update((T) t));
        menuActions.put(DELETE_CODE, w -> {
            database.delete((T) w);
            container.remove(w);
            notifyDataSetChanged();
        });

        container = (ArrayList<T>) database.readAll();
    }

    public void add(T element){
        container.add(element);
        notifyDataSetChanged();
    }

    @Override
    public int getGroupCount() {
        return container.size();
    }

    @Override
    public Object getGroup(int groupPosition) {
        return container.get(groupPosition);
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return null;
    }

    @Override
    public boolean hasStableIds() {
        return false;
    }

    @Override
    public long getGroupId(int groupPosition) {
        return 0;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return 0;
    }

    @Override
    public View getGroupView(int groupPosition, boolean isExpanded, View convertView, ViewGroup parent) {
        View view = getInflater().inflate(R.layout.manager_group_item, parent, false);

        TextView title = ButterKnife.findById(view, R.id.name);
        title.setText(container.get(groupPosition).getName());

        ImageView option = ButterKnife.findById(view, R.id.options_button);
        option.setOnClickListener(v -> showPopupMenu(v, container.get(groupPosition)));

        return view;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return false;
    }

    private void showPopupMenu(View anchor, T selected){
        PopupMenu menu = new PopupMenu(context, anchor);
        menu.getMenu().add(Menu.NONE, EDIT_CODE, Menu.NONE, context.getString(R.string.action_edit));
        menu.getMenu().add(Menu.NONE, DELETE_CODE, Menu.NONE, context.getString(R.string.action_delete));

        menu.setOnMenuItemClickListener(item -> {
            menuActions.get(item.getItemId()).run(selected);

            return true;
        });

        menu.show();
    }

    protected LayoutInflater getInflater() {
        return inflater;
    }

    protected ArrayList<T> getContainer() {
        return container;
    }

    protected CRUD<T> getDatabase() {
        return database;
    }

    protected static final int EDIT_CODE = 1;
    protected static final int DELETE_CODE = 2;

    @FunctionalInterface
    protected interface OnMenuItemSelectedAction<T>{

        void run(T element);
    }


}

File: app/src/main/java/com/bearcave/passageplanning/base/BaseManagerAdapter.kt
package com.bearcave.passageplanning.base

import android.content.Context
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.ImageView
import android.widget.PopupMenu
import android.widget.TextView
import butterknife.ButterKnife
import com.bearcave.passageplanning.R
import com.bearcave.passageplanning.base.database.withcustomkey.CRUDWithCustomKey
import com.bearcave.passageplanning.base.database.withcustomkey.DatabaseElementWithCustomKey
import com.bearcave.passageplanning.tasks.BackgroundTask
import java.util.*

/**
 * All ExpandableListViews used in BasePoorManager should extend this class.
 *
 * @see BasePoorManagerFragment
 * @author Michał Wąsowicz
 * @version 1.0
 * @since 20.05.17
 */
abstract class BaseManagerAdapter<Dao : DatabaseElementWithCustomKey<T>, T>
    (parent: BasePoorManagerFragment<*, *>, protected val context: Context)
    : BaseExpandableListAdapter() {

    /**
     * This container has all DAOs shown in ExpandableListView.
     */
    protected var container = ArrayList<Dao>()


    protected val inflater: LayoutInflater
        get() = LayoutInflater.from(context)

    @Suppress("UNCHECKED_CAST")
    protected val database: CRUDWithCustomKey<Dao, T> = parent as CRUDWithCustomKey<Dao, T>
    private val commands = ArrayList<Command>()


    init {
        AdapterTask().execute({
            container = database.readAll() as ArrayList<Dao>
        })
    }

    /**
     * @param element - send from external class to be added in list view.
     */
    fun add(element: Dao) {
        container.add(element)
        notifyDataSetChanged()
    }

    /**
     * @param toBeUpdated - sent from external class to be updated in list view.
     */
    fun update(toBeUpdated: Dao) {
        for ((i, element) in container.withIndex()) {
            if (element.id == toBeUpdated.id) {
                container[ i ] = toBeUpdated
                notifyDataSetChanged()
                return
            }
        }
    }

    override fun getGroupCount() = container.size

    override fun getGroup(groupPosition: Int): Any = container[groupPosition]

    override fun getChild(groupPosition: Int, childPosition: Int): Any? = null

    override fun hasStableIds() = false

    override fun getGroupId(groupPosition: Int): Long = 0

    override fun getChildId(groupPosition: Int, childPosition: Int): Long = 0

    override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convertView: View?, parent: ViewGroup): View {
        val view = convertView ?: inflater.inflate(R.layout.manager_group_item, parent, false)

        val title = ButterKnife.findById<TextView>(view, R.id.name_placeholder)
        title.text = container[groupPosition].name

        val option = ButterKnife.findById<ImageView>(view, R.id.options_button)
        option.setOnClickListener { v -> showPopupMenu(v, container[groupPosition]) }

        return view
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int) = false

    private fun showPopupMenu(anchor: View, selected: Dao) {
        val menu = PopupMenu(context, anchor)
        for (command in commands){
              menu.menu.add(Menu.NONE, command.i, Menu.NONE, command.name)
        }

        menu.setOnMenuItemClickListener { item ->
            commands[item.itemId].command(selected)
            true
        }

        menu.show()
    }

    var INDEX = 0 // used only in addOption for options indexing.


    protected fun addOption(title: String, option: (Dao)->Unit) {
        commands.add(Command(INDEX++, title, option))
    }


    private inner class Command(val i: Int, val name: String, val command: (Dao)->Unit)

    private inner class AdapterTask : BackgroundTask(context) {
        override fun onPostExecute(result: Int) {
            super.onPostExecute(result)
            notifyDataSetChanged()
        }
    }
}
